### ISSUE TITLE
multiple lua scripts bug?

### ISSUE THREAD

**Issue by basvandervlies** [stanford-rc/slurm-spank-lua#14]
I have a question can we have multiple lua scripts in the `lua.d` directory. From our testing only 1 script is executed. We have 2 lua scripts. But we only see the fist one being executed twice:
 * 1jaap.kua
```
myname = "jaap"

function slurm_spank_task_init_privileged (spank)
    SPANK.log_info("jaap")
    SPANK.log_info("bas en jaap")
    return 0
end
```
 * bas.lua
```
myname = "bas"

function slurm_spank_task_init_privileged (spank)
    SPANK.log_info("bas")
    return 0
end
```

We see in the in the slurmd log:
````
[2022-02-08T16:53:36.750] [8737421.0] jaap
[2022-02-08T16:53:36.750] [8737421.0] bas en jaap
[2022-02-08T16:53:36.750] [8737421.0] jaap
[2022-02-08T16:53:36.750] [8737421.0] bas en jaap
```

so it sees two scripts but only 1 is executed twice!

**Comment by kcgthb** [comment]
Hi @basvandervlies 
I'm afraid can't reproduce this. With the following 2 files in `/etc/slurm/lua.d`:
```
$  ls -al test*
-rw-r--r-- 1 root root 95 Feb  8 10:35 test1.lua
-rw-r--r-- 1 root root 95 Feb  8 10:35 test2.lua
$ cat test1.lua
function slurm_spank_task_init_privileged (spank)
    SPANK.log_info("test1")
    return 0
end
$ cat test2.lua
function slurm_spank_task_init_privileged (spank)
    SPANK.log_info("test2")
    return 0
end
```

I get the following log when a job starts:
```
Feb 08 10:35:52 sh02-01n59.int slurmstepd[9295]: in _window_manager
Feb 08 10:35:52 sh02-01n59.int slurmstepd[9326]: test2
Feb 08 10:35:52 sh02-01n59.int slurmstepd[9326]: test1
```

So it looks like both Lua scripts get executed, once each.

Are you positive both scripts have the correct `.lua` extension and are accessible and readable? That may just be a typo here, but I see your first script was named `1jaap.kua` instead of `1jaap.lua`.

**Comment by basvandervlies** [comment]
Yes I am sure both scripts run. If I rename `1jaap.lua` to `jaap.lua`.  Then `bas.lua` runs twice. Which Lua version do you use.  I suspect 5.1. We are using 5.3

**Comment by kcgthb** [comment]
Yes, we're using Lua 5.1, indeed.

**Comment by basvandervlies** [comment]
lua greater then 5.1 the code for this has been changed. What I am seeing is that the plugin sees 2 lua scrips and calls 2 times the function but it is the function of the scripti that is first loaded that is why I was playing with the `1` prefix. Here is what I see in debug mode:  (slurm_spank_job_prolog added to both lua scripts)
```
spank-prolog: spank/lua: Loaded 2 plugins in this context
spank-prolog: jaap job_prolog
spank-prolog: jaap job_prolog
```

so it is called twice with the same prolog function. So code greater the 5.1 some goes wrong in the assigment of the spank functions. If I add a third script it the same functions is called 3 times.

**Comment by kcgthb** [comment]
Thanks for the details.
I'd recommend reporting this to the original upstream project at https://github.com/grondo/slurm-spank-plugins/issues, since this one is mainly a repackaging of the original SPANK Lua plugin, without the other plugins, but without any major code modification either.

**Comment by basvandervlies** [comment]
@kcgthb the port to 5.2/5.3 has been done in this fork and not in the upstream one.  The upstream supports only 5.1 and I think they won't fix it. Maybe @treydock knows what the problem is?  He has done the port for 5.3.

I am using lua 5.2. The problem is for 5.2/5.3

**Comment by kcgthb** [comment]
I gave it a try with Lua 5.3, and I can indeed reproduce the behavior you described.
It's even funnier than that, since dropping a 3rd test script in the `lua.d` directory makes `slurmstepd` log the same thing 3 times:
```
$ find /etc/slurm/lua.d/ -name 'test*'
/etc/slurm/lua.d/test1.lua
/etc/slurm/lua.d/test2.lua
/etc/slurm/lua.d/test3.lua
```
```
$ journalctl -u slurmd | grep test
Feb 09 09:36:48 sh02-01n59.int slurmstepd[7573]: test1
Feb 09 09:36:48 sh02-01n59.int slurmstepd[7573]: test1
Feb 09 09:36:48 sh02-01n59.int slurmstepd[7573]: test1
```

To make sure that it was not just a logging issue, I modified the test lua scripts to create a file and append a script-specific string (`test1.lua` appends `test1` to `/tmp/test1.out`, `test2.lua` appends `test2` to `/tmp/test2.lua`, etc...) and sure thing, only `/tmp/test1.lua` is created, and it contains 3 `test1` lines. So that confirms that the 1st script is executed 3 times.

I suspect this may be related to some changes in the way Lua 5.2+ handles global variables maybe? But I'm afraid I don't know Lua's internals enough to identify the issue. Maybe @treydock will know.

**Comment by treydock** [comment]
My changes to support Lua for RHEL8 were just enough to get it to compile, and that was a struggle given how little I know about Lua internals and their C interface.  I won't have a RHEL8 Slurm cluster for a few more months but if I run into same issue I'd be happy to help debug and see what I can find.  I just don't have any RHEL8 systems to test with at this time ( so still on Lua 5.1).

**Comment by kcgthb** [comment]
Hi @basvandervlies 
Thanks to @martinetd, this issue should now be fixed in v0.44. Please feel free to give it a try and let me know how it goes!

**Comment by basvandervlies** [comment]
Thanks @kcgthb and @martinetd I was on the same route ;-) 

I found it in another source and want to ask the original autrhor for consent.  https://github.com/basvandervlies/slurm-spank-lua/commit/5a95c88862852945d852b6261eb48e31a71fbf45
The code above worked  for lua 5.1 and lua 5.3. 

I will test the new code when I am back from holidays. Great We have found an solution!
